plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/Attribute.java

```diff
@@ -75,6 +80,69 @@ static void writeTo(final StreamOutput out, final Attribute attribute) throws IO
         out.writeString(attribute.toString());
     }
 
+    /**
+     * Write Attribute value to a StreamOutput
+     * @param out the StreamOutput to write
+     * @param attributeValue the Attribute value to write
+     */
+    public static void writeValueTo(StreamOutput out, Object attributeValue) throws IOException {
```

Comment:
We may not need this if this is only for serializing `source`. I just did some experiments by removing the toString in the listener when saving the searchqueryrecord, it seems to be working since SearchSourceBuilder also implements ToXContentObject. The top n queries results I got:
```
 {
      "timestamp" : 1719384404639,
      "labels" : { },
      "source" : {
        "size" : 20,
        "query" : {
          "bool" : {
            "must" : [
              {
                "match_phrase" : {
                  "message" : {
                    "query" : "document",
                    "slop" : 0,
                    "zero_terms_query" : "NONE",
                    "boost" : 1.0
                  }
                }
              },
              {
                "match" : {
                  "user.id" : {
                    "query" : "cyji",
                    "operator" : "OR",
                    "prefix_length" : 0,
                    "max_expansions" : 50,
                    "fuzzy_transpositions" : true,
                    "lenient" : false,
                    "zero_terms_query" : "NONE",
                    "auto_generate_synonyms_phrase_query" : true,
                    "boost" : 1.0
                  }
                }
              }
            ],
            "adjust_pure_negative" : true,
            "boost" : 1.0
          }
        }
      },
      "search_type" : "query_then_fetch",
...
```

Reply:
This was failing without custom handling for writing and reading SearchSourceBuilder with write and read errors. Had to do the custom handling to overcome the errors.

Reply:
Mainly the serialize then deserialize and compare test was failing

Reply:
If the functions are only used in tests, I think it make more sense to write helper functions instead of adding functions in the Attribute code

Reply:
Looks like most of these are used in the actual code path too

Reply:
It is only used in the `writeTo` function, which is for serializing `SearchQueryRecord`, as I mentioned in the comment, we don't need to do 
```
out.writeMap(
            attributes,
            (stream, attribute) -> Attribute.writeTo(out, attribute),
            (stream, attributeValue) -> Attribute.writeValueTo(out, attributeValue)
        );
```
, 
just 
```
out.writeMap(attributes, (stream, attribute) -> Attribute.writeTo(out, attribute), StreamOutput::writeGenericValue);
```
is enough for serialization based on my experiments. Otherwise could you give an example on what could go wrong if we don't write the Attribute map based on type?

Reply:
[Here](https://build.ci.opensearch.org/job/gradle-check/41922/testReport/junit/org.opensearch.plugin.insights.rules.action.top_queries/TopQueriesResponseTests/testSerialize/) is the error we are getting when we use the above approach recommended by you.

Reply:
I made the same change on master branch
```
attributes.put(Attribute.SOURCE, request.source().toString(FORMAT_PARAMS));
->
attributes.put(Attribute.SOURCE, request.source());
```
Those tests are passing. I think some changes in this PR is breaking the test, could you double check?

Reply:
Did more research on this. This looks like a urgent bug to me. I have included your code in this PR https://github.com/opensearch-project/OpenSearch/pull/14681/files. Let's get it in ASAP. Thanks for the finding!